### PR TITLE
WIP Scrolling

### PR DESCRIPTION
**Pull request by joshka**
This is just a work in progress PR. Feel free to take ideas and use them in other PRs (e.g. #228)

Putting this here just for feedback on the idea and to communicate some of the things more concretely from #174

### CODE REVIEW

File: src/widgets/scroll.rs [hunk 1]
```diff
@@ -0,0 +1,265 @@
...
+}
+
+#[derive(Clone, Copy, Debug, Default, Eq, PartialEq)]
+enum ScrollDirection {
+    #[default]
+    Up,
+    Down,
+    Left,
+    Right,
+}
```
Comment (kdheepak) [inline src/widgets/scroll.rs:17]: I'm curious why `ScrollDirection` isn't just `Vertical` and `Horizontal`?

File: src/widgets/scroll.rs [hunk 2]
```diff
@@ -0,0 +1,296 @@
...
+    }
+    fn first_item(&mut self) { self.scroll_item_index = 0; }
+    fn last_item(&mut self) { self.scroll_item_index = self.item_count.saturating_sub(1) }
+
+    fn scroll_up(&mut self) { self.scroll_position.y = self.scroll_position.y.saturating_sub(1) }
+    fn scroll_down(&mut self) { self.scroll_position.y = self.scroll_position.y.saturating_add(1) }
+    fn scroll_left(&mut self) { self.scroll_position.x = self.scroll_position.x.saturating_sub(1) }
+    fn scroll_right(&mut self) { self.scroll_position.x = self.scroll_position.x.saturating_add(1) }
+
+    fn scroll_direction(&mut self, direction: ScrollDirection) {
```
[the remark this answers is not in the document]
Comment (joshka) [reply src/widgets/scroll.rs:48]: This is a good idea.

File: src/widgets/scroll.rs [hunk 3]
```diff
@@ -0,0 +1,296 @@
...
+#![allow(unreachable_code)]
+
+use crate::{buffer::Buffer, layout::Rect, style::Style};
+
+use super::StatefulWidget;
+
+#[derive(Clone, Copy, Debug, Default, Eq, PartialEq)]
+struct ScrollPosition {
+    x: u16,
+    y: u16,
```
Comment (hasezoey) [inline src/widgets/scroll.rs:11]: maybe change this to be `usize` to allow content to be bigger than draw area

**Comment by joshka** [comment]
Closing out as a stale PoC

**Pull request closed without merging**